### ISSUE TITLE
Effort Tracking concerns

### ISSUE THREAD

**Issue by CaraesNaur** [flyspray/flyspray#898]
### Concurrent time per user

Effort tracking (in `effort::startTracking()`) only checks whether a user is already tracking time against the current task.

It does not disallow a user having multiple concurrent effort trackers active each on *different* tasks.  The `effort` class is unaware of projects, only user and task.

Perhaps there should be a project pref for allowing concurrent time tracking?

- Note: "Clear Tracking" deletes *all* open effort entries for the task, which leads me to believe the code otherwise assumes there will only be one per `task+user`.

### Manual Effort Entry

When entering effort manually, the current time is used for both `start_time` and `end_time`, yet the amount of effort is stored as expected.  This doesn't make much sense.

Solution: Manual Effort entry requires inputs for `start_time` and `end_time`.  It looks like JSCalendar can include time, but not *only* time.  Would have to test for Time Zone issues.

Manual Entry should verify that `start_time` and `end_time` are both in the past.  We could decide that if `end_time` is blank (and no other open entry for the `task+user` exists), the entry would be created in the active state, as if "Start Tracking" had been clicked at `$start_time`.  Probably needs additional logic to limit duration.

### Effort class should be capable of reporting whether effort is being tracked

`startTracking()` executes two queries:

* Count rows in `{effort}` by `task_id` and `user_id`
* Create a new row in `{effort}`

The first query would be useful to expose on its own, such as to hide the "Start Tracking" button on the Details Effort tab if tracking is already active.

> I have already done this in my #887 branch.  `startTracking()` now calls the new function `countActiveTracking()`, which returns a count of open `{effort}` rows on the task by the user.

### "Start Tracking" Should Include Description

Otherwise there is no means of populating that field.

### Effort Tracking continues indefinitely

The only way effort tracking gets terminated is by *the user* clicking "Stop Tracking".

If an effort item is forgotten about, or the user somehow becomes inactive (account removed, IRL happenings, etc) the tracking will continue forever.

As far as I can tell, no other user (Admin, PM) is currently capable of stopping effort tracking.  The Effort Tracking has a lot of room for expansion, especially for Admin/PM/reporting.

If Flyspray had automation capability, one use for that would be to auto-expire open Effort Tracking after 1 man-day.  Which leads to...

### Default Man-Day

Default man-day should be a reasonable real-world value such as `8:00`.  `0:00` should be disallowed, as it can lead to many calculation issues.

### Make Active Effort Tracking Easily Accessible

Add an always-visible item in `#menu-list` or `#pm-menu` when effort tracking is active.

`#pm-menu` seems more logical because PM requests appear there, however this item would need to be project agnostic, showing active effort in all projects.  The item would open a menu of links to tasks with active effort tracking, or simply link to an Effort Tracking tab in `myprofile`.

---

That's what I stumbled into from tweaking the Effort tab UI in `details`.

**Comment by peterdd** [comment]
Well, effort tracking is not a first citizen of Flyspray features. What exists now is just appended to Flyspray on request/wishes and because it was quite easily to add to the existing Flyspray code base.

It allows a Flyspray user having tasks 'running'/tracking in parallel (intentional) and does not enforce anything.

- If I do not halluzinate, there was once a scenario where (mechanical/production) machines were used as task 'taker'. So such multitask-machines could do multiple tasks concurrently and are independent of human daily work or times.
- Other scenarios where effort could be split across multiple projects (e.g. customer projects) like inventing a module/extension for multiple customers and split the costs by billing it across them. (e.g. development costs are higher then what one customers would pay.)
- Flyspray user accounts could also be teams (of workers or day laborers) or machines, so things like 'man-day' hours differ depending of projects and organizations. Think of a building team where the team leader inputs the hours done for a task or like that.

So some try/tried to use Flyspray also as source for billing work hours to customers, but yeah it is really basic.

> "Start Tracking" Should Include Description
> Otherwise there is no means of populating that field.

This optional description was added recently (https://github.com/flyspray/flyspray/commit/ed5870a39522ee92eecd2ec39988d39c5437d720) on request, so for manual hour entries it is possible to add a description to give details for insert billing hours of a tasks like:

```
1:00 travel
2:00 plumbing
```

Maybe when stopping giving optional details could be added first (currently mytimers.php and in task details) .

Also add/edit a description for an existing finished effort entry maybe could be added.
(in what situations allowed - (config, time..), who is allowed, ..)
